[FLINK-33139] Add Prometheus Sink Table API

darenwkt opened the pull request:
<!--
*Thank you for contributing to the Apache Flink Prometheus Connector - we are happy that you want to help us improve our Flink connectors. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

## Contribution Checklist

- The name of the pull request should correspond to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
- Commits should be in the form of "[FLINK-XXXX][Connectors / Prometheus] Title of the pull request", where [FLINK-XXXX] should be replaced by the actual issue number.
- Each pull request should only have one JIRA issue.
- Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.
-->

## Purpose of the change

https://issues.apache.org/jira/browse/FLINK-33139 Implements Table API for Prometheus Sink

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change added tests and can be verified as follows:
- Added unit tests for RowDataElementConverter and DynamicSinkTableFactory
- Manually verified by running the Prometheus Sink TableAPI to submit data to Amazon Managed Prometheus
![Screenshot 2025-03-31 at 22 00 00](https://github.com/user-attachments/assets/cb7c3cb1-3ec8-4fc3-a46e-a4d848a33c50)

  - SQL Code ran for the test:
```
CREATE TABLE PromTable (
  `my_metric_name` BIGINT,
  `my_label` BIGINT,
  `sample_value` BIGINT,
  `sample_ts` TIMESTAMP(3)
)
WITH (
  'connector' = 'prometheus',
  'aws.region' = 'us-east-1',
  'metric.request-signer' = 'amp',
  'metric.endpoint-url' = 'https://aps-workspaces.us-east-1.amazonaws.com/workspaces/redacted/api/v1/remote_write',
  'metric.name' = 'my_metric_name',
  'metric.label.keys' = 'my_label',
  'metric.sample.key' = 'sample_value',
  'metric.sample.timestamp' = 'sample_ts',
  'sink.batch.max-size' = '2',
  'sink.flush-buffer.size' = '1'
);

CREATE TABLE DataGen (
  `my_metric_name` STRING,
  `my_label` STRING,
  `sample_value` BIGINT,
  `sample_ts` TIMESTAMP(3)
) WITH (
  'connector' = 'datagen',
  'rows-per-second' = '50',
  'fields.my_metric_name.kind' = 'random',
  'fields.my_metric_name.values' = '["my_metric_1", "my_metric_2"]',
  'fields.my_label.kind' = 'random',
  'fields.my_label.values' = '["my_label_1", "my_label_2"]'
);

INSERT INTO PromTable SELECT my_metric_name, my_label, sample_value, sample_ts FROM DataGen;
```

- Manually verified that existing Prometheus Sink DataStream API is still working
<img width="399" alt="Screenshot 2025-04-02 at 09 38 23" src="https://github.com/user-attachments/assets/20af1d1c-20c3-46ed-8306-32756c3b0f8a" />

## Significant changes
*(Please check any boxes [x] if the answer is "yes". You can first publish the PR and check them afterwards, for convenience.)*
- [yes] Dependencies have been added or upgraded
- [no] Public API has been changed (Public API is any class annotated with `@Public(Evolving)`)
- [no] Serializers have been changed
- [yes] New feature has been introduced
    - If yes, how is this documented? (JavaDocs)

leekeiabstraction commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSink.java:
```diff
@@ -37,7 +37,7 @@
 /** Sink implementation accepting {@link PrometheusTimeSeries} as inputs. */
 @PublicEvolving
-public class PrometheusSink extends AsyncSinkBase<PrometheusTimeSeries, Types.TimeSeries> {
+public class PrometheusSink<InputT> extends AsyncSinkBase<InputT, Types.TimeSeries> {
```
Is this a breaking change? Would current code which has reference to PrometheusSink compile successfully?

leekeiabstraction replied:
Also on name of type parameter, should we follow convention of using capital letter?

darenwkt replied:
> Is this a breaking change? Would current code which has reference to PrometheusSink compile successfully?

Yes, this is a breaking change, current code will not compile successfully as it is missing the InputType specification, I have updated the PR description with a breaking change summary to explain this in more details

> Also on name of type parameter, should we follow convention of using capital letter?

Thanks, this is a good shout, have renamed from "InputType" to "IN"

leekeiabstraction commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkBuilder.java:
```diff
@@ -30,9 +30,8 @@
-        extends AsyncSinkBaseBuilder<
-                PrometheusTimeSeries, Types.TimeSeries, PrometheusSinkBuilder> {
+public class PrometheusSinkBuilder<InputT>
+        extends AsyncSinkBaseBuilder<InputT, Types.TimeSeries, PrometheusSinkBuilder<InputT>> {
```
Similar comment on breaking change and also type parameter.

darenwkt replied:
Addressed in PR description's breaking change summary

leekeiabstraction commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusTimeSeriesConverter.java:
```diff
@@ -22,42 +22,23 @@
-        implements ElementConverter<PrometheusTimeSeries, Types.TimeSeries> {
-
-    private static final String METRIC_NAME_LABEL_NAME = "__name__";
+public class PrometheusTimeSeriesConverter<InputT>
+        implements ElementConverter<InputT, Types.TimeSeries> {
```
Similar comment on breaking change.

darenwkt replied:
Addressed in PR description's breaking change summary

foxus commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/PrometheusConfig.java:
```diff
@@ -0,0 +1,75 @@
+        return options.get(METRIC_NAME);
+    }
+
+    public List<String> getLabelKeys() {
```
Are `labelKeys` unique? If so, consider returning a set rather than a list to enforce uniqueness, alternatively consider calling unique after the split. This will protect against a user supplied configuration including duplicates and also ensure that a serialized Set is identical regardless of the order of keys provided.

darenwkt replied:
Thanks, this is a good shout, I have updated it to a set

foxus commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/PrometheusConnectorOption.java:
```diff
@@ -0,0 +1,75 @@
+                    .noDefaultValue()
+                    .withDescription("Prometheus metric name.");
+
+    public static final ConfigOption<String> METRIC_LABEL_KEYS =
```
Related to previous comment, `ConfigOption` has native support for CSVs -

```
 ConfigOption<String> METRIC_LABEL_KEYS = ConfigOptions
     .key("metric.label.keys")
     .stringType()
     .asList()
     .noDefaultValue()
     .withDescription("Prometheus metric label key name.");
```

You could then eliminate the streaming and hardcoded delimiter work above.

foxus replied:
I would also consider rewording the description to make it clear there can be more than one key name provided.

darenwkt replied:
> Related to previous comment, ConfigOption has native support for CSVs

Thanks, this is a great suggestion, have updated PR with it

> I would also consider rewording the description to make it clear there can be more than one key name provided.

Updated description

foxus commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/PrometheusDynamicSink.java:
```diff
@@ -0,0 +1,193 @@
+                        .setPrometheusRemoteWriteUrl(prometheusConfig.getRemoteWriteEndpointUrl());
+
+        Optional.ofNullable(prometheusConfig.getRequestSignerIdentifier())
+                .ifPresent(region -> builder.setRequestSigner(getRequestSigner(prometheusConfig)));
```
I'm uncertain why `region` is being created within the context of a request signer identifier only not to be used in the subsequent call. Would this be simpler to leave what's required from the `prometheusConfig` to the `getRequestSigner` method which will return null if there isn't one and only call `builder.setRequestSigner` if it isn't?

darenwkt replied:
I agree, on top of your point, moving it to getRequestSigner is better here especially since we want to separate any Amazon config away from flink-connector-prometheus package. Updated and added unit test for this

foxus commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/PrometheusDynamicSink.java:
```diff
@@ -0,0 +1,193 @@
+        Iterator<PrometheusDynamicRequestSignerFactory> factories = loader.iterator();
+        String requestSignerIdentifier = prometheusConfig.getRequestSignerIdentifier();
+
+        while (true) {
+            try {
+                if (!factories.hasNext()) {
+                    break;
+                }
+
+                PrometheusDynamicRequestSignerFactory factory = factories.next();
```
Would this be simpler as a for loop as well as incorporating line 129?

```
for(Iterator<PrometheusDynamicRequestSignerFactory> factories = loader.iterator(); factories.hasNext(); ) {
   PrometheusDynamicRequestSignerFactory factory = factories.next();
```

darenwkt replied:
I have converted it to a Stream to make this more readable

foxus commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/PrometheusDynamicSinkFactory.java:
```diff
@@ -0,0 +1,165 @@
+                                        () ->
+                                                new ValidationException(
+                                                        String.format(
+                                                                "%s not found in fields: %s",
```
Could we consider something along the lines of `%s is not a valid Prometheus metric config key, valid keys are: %s`, this message is exposing the implementation of the validation rather than what the user should do to fix it.

darenwkt replied:
Adopted suggestion

foxus commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/RowDataToPrometheusTimeSeriesConverter.java:
```diff
@@ -0,0 +1,111 @@
+            } else if (fieldName.equals(prometheusConfig.getMetricSampleTimestamp())) {
+                sampleTimestamp = fieldValue.getLongValue();
+            } else {
+                throw new IllegalArgumentException("Unsupported field: " + fieldName);
```
Could we consider providing the list of supported fields in this exception message?

darenwkt replied:
Adopted suggestion

foxus reviewed (changes requested):
Many thanks for your efforts here, I've added a few comments.

darenwkt commented:
Thanks @leekeiabstraction @foxus for the review, I have addressed the comments and updated the PR, please take a look

foxus commented on flink-connector-prometheus-request-signer-amp/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSignerFactory.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private String getAwsRegionFromEndpointUrl(String endpointUrl) {
+        String regex = "aps-workspaces\\.([^.]+)\\.amazonaws\\.com";
```
I believe this is too prescriptive - there are FIPS endpoints which this will not match - consider reviewing https://docs.aws.amazon.com/general/latest/gr/prometheus-service.html and making this regex less specific.

darenwkt replied:
Thanks, this is a good catch, have updated this

foxus commented on flink-connector-prometheus-request-signer-amp/src/test/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSignerFactoryTest.java:
```diff
@@ -0,0 +1,55 @@
+        assertEquals(factory.requestSignerIdentifer(), "amazon-managed-prometheus");
+    }
+
+    @Test
```
Consider using a `DataProvider` to test all the endpoints listed in https://docs.aws.amazon.com/general/latest/gr/prometheus-service.html.

darenwkt replied:
Thanks, yes I have extracted all the endpoints from the doc and used them in the unit test

foxus reviewed (changes requested):
Thanks for the updates, loving the work to migrate the AWS specific Prometheus behaviour out. I've added additional comments on the updates.

darenwkt commented:
> Thanks for the updates, loving the work to migrate the AWS specific Prometheus behaviour out. I've added additional comments on the updates.

Thanks for the review, have updated PR to address them, please take another look

foxus reviewed (approved):
LGTM - thank you for the timely updates!

nicusX commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/PrometheusConnectorOption.java:
```diff
@@ -0,0 +1,67 @@
+    // -----------------------------------------------------------------------------------------
+
+    public static final ConfigOption<String> METRIC_NAME =
+            ConfigOptions.key("metric.name")
```
I am not sure I understand the interface.
Is this configuration specifying the metric name as a constant string, or the field of the table that contains the metric name?

In the example, it seems to be the latter.
It should be, because in a realistic use each record can be from a different metric.

However, looking at the example
```
CREATE TABLE PromTable (
  `my_metric_name` BIGINT,
  `my_label` BIGINT,
  ...
)
WITH (
  'connector' = 'prometheus',
  'metric.name' = 'my_metric_name',
  'metric.label.keys' = 'my_label'
  ....
);
```
but the field is a BIGINT, while it should be a VARCHAR.

Also, labels are key/value (String/String) pairs. 
Each record (each measurement) can have a different set of Labels (Set of K/V) where both key and value may change

darenwkt replied:
Thanks Lorenzo for the review and apology for the misleading example as there were typo in there, I have updated description and below with an accurate example:

In the example below, a table is created where the metricName, labels are STRING and there are 2 labels (String/String) pairs used as you have correctly pointed out.

```
Flink SQL> CREATE TABLE PromTable (
>   `my_metric_name` STRING,
>   `my_label_1` STRING,
>   `my_label_2` STRING,
>   `sample_value` BIGINT,
>   `sample_ts` TIMESTAMP(3)
> )
> WITH (
>   'connector' = 'prometheus',
>   'metric.endpoint-url' = 'https://aps-workspaces.us-east-1.amazonaws.com/workspaces/redacted/api/v1/remote_write',
>   'metric.name' = 'my_metric_name',
>   'metric.label.keys' = '[my_label_1,my_label_2]',
>   'metric.sample.key' = 'sample_value',
>   'metric.sample.timestamp' = 'sample_ts'
> );
[INFO] Execute statement succeed.
```

Next, a row of sample data is inserted into the table where metricName="sample_name", and the label pairs are ("my_label_1":"sample_label_1", "my_label_2":"sample_label_2")

```
Flink SQL> INSERT INTO PromTable VALUES ('sample_name', 'sample_label_1', 'sample_label_2', 123, TIMESTAMP '2025-04-30 10:00:00.000');
[INFO] Submitting SQL update statement to the cluster...
[INFO] SQL update statement has been successfully submitted to the cluster:
Job ID: 596c86fb5485ddd3c1a3594e3c837d02
```

nicusX commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSink.java:
```diff
@@ -49,11 +49,11 @@ public class PrometheusSink extends AsyncSinkBase<PrometheusTimeSeries, Types.Ti
     @SuppressWarnings("checkstyle:RegexpSingleline")
     protected PrometheusSink(
-            ElementConverter<PrometheusTimeSeries, Types.TimeSeries> elementConverter,
+            ElementConverter<IN, Types.TimeSeries> elementConverter,
```
Is this change required for supporting Table API or just a matter of generalizing the API for something we are not really generalizing? The DataStream sink only supports PrometheusTimeSeries as input

Breaking the public interface not only forces a major version, but because it forces existing users to change their code

darenwkt replied:
I have taken a look and have found a way to support TableAPI without breaking change.

In summary, the @PublicEvolving API here are:
- PrometheusSink
- PrometheusSinkBuilder
- PrometheusTimeSeriesConverter

The challenge here is that the above 3 classes need to support RowData as inputType for TableAPI implementation.

There are two ways to solve this:
- Option 1 - Make all 3 classes take a generic input type (Causes breaking change)
- Option 2 - Introduce a Base class with generic input type for all 3 classes which can now inherit it. (No breaking change)

 From your comment, I agree that breaking change is unnecessary and I have updated the PR to go for Option 2.

nicusX commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSink.java:
```diff
@@ -49,11 +49,11 @@ public class PrometheusSink extends AsyncSinkBase<PrometheusTimeSeries, Types.Ti
             int maxBufferedRequests,
             int maxBatchSizeInSamples,
-            int maxRecordSizeInSamples,
+            long maxRecordSizeInSamples,
```
Why changing this? The batch size will never be > Integer.MAX_VALUE samples

darenwkt replied:
It's because I noticed that `maxBatchSizeInBytes` attribute in `AsyncSinkBase` class is Long and we have casted it from int to long in the super constructor of PrometheusSink. So, I thought we should match it to the super class, but I realised this is a breaking change that is not needed, so I have reverted it.

nicusX commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusTimeSeriesConverter.java:
```diff
@@ -22,42 +22,22 @@
-                            .setValue(sample.getValue())
-                            .setTimestamp(sample.getTimestamp())
-                            .build());
+    public Types.TimeSeries apply(IN element, SinkWriter.Context context) {
```
As my previous point: if we still support PrometheusTimeSeries only, is it worth breaking the public API?

darenwkt replied:
Answered in previous point: we will not introduce breaking change

nicusX reviewed (changes requested):
Thanks for the PR
I am not sure I fully understand the Table API interface and whether it matches the use case this connector is supposed to be used for. I left some specific comments about the API. Would you please clarify?

Also, I am concerned of breaking the public interface of DataStream API. Is that required?

darenwkt commented:
> Thanks for the PR I am not sure I fully understand the Table API interface and whether it matches the use case this connector is supposed to be used for. I left some specific comments about the API. Would you please clarify?
> 
> Also, I am concerned of breaking the public interface of DataStream API. Is that required?

Thanks Lorenzo for the review, I have addressed comments on the API use case and example and I hope that it clarifies it better now.

I have also found a way to support TableAPI without breaking changes, appreciate if you can take another look at the updated PR, thank you

georgelza commented:
hi guys...
is there a jar file available for this connector... looks like exactly what i've been looking for.
curious, looking at the Create table, don't see mention of username/password... credentials... just asking, 
G

georgelza commented:
excuse the noob question, so I did the git clone, and the build, the project folder has a couple of directories with target sub folders. which jar from which sub directory do i copy to my flink cluster/lib folder for the above to work ?

darenwkt commented:
Hi @georgelza.

> hi guys... is there a jar file available for this connector... looks like exactly what i've been looking for. curious, looking at the Create table, don't see mention of username/password... credentials... just asking, G

You can use request signer to provide authentication/credentials, you can see the Flink doc for more info on this https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/datastream/prometheus/#request-signer

> excuse the noob question, so I did the git clone, and the build, the project folder has a couple of directories with target sub folders. which jar from which sub directory do i copy to my flink cluster/lib folder for the above to work ?

You should not need to build it, the jar is available in maven https://mvnrepository.com/artifact/org.apache.flink/flink-connector-prometheus/1.0.0-1.20

georgelza commented:
> Hi @georgelza.
> 
> > hi guys... is there a jar file available for this connector... looks like exactly what i've been looking for. curious, looking at the Create table, don't see mention of username/password... credentials... just asking, G
> 
> You can use request signer to provide authentication/credentials, you can see the Flink doc for more info on this https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/datastream/prometheus/#request-signer
> 
> > excuse the noob question, so I did the git clone, and the build, the project folder has a couple of directories with target sub folders. which jar from which sub directory do i copy to my flink cluster/lib folder for the above to work ?
> 
> You should not need to build it, the jar is available in maven https://mvnrepository.com/artifact/org.apache.flink/flink-connector-prometheus/1.0.0-1.20

thanks, will pull the jar from maven using the above link, just want to double check does the jar thats available include the create table ... (connector=prometheus...) functionality... as per above. I'm asking as I build as per the instructions and even though the jar file was in the correct place it seems flink was not able to find the prometheus connector functionality.
would appreciate if you can ping me privately to discuss further. georgelza@gmail.com
G

georgelza commented:
@darenwkt No luck. does it make sense to to this rather off here... 

`Flink SQL> CREATE TABLE hive_catalog.prometheus.PromTable (
>    sensorid         STRING,
>    siteId           STRING,
>    deviceId         STRING,
>    measurement      BIGINT,
>    ts               TIMESTAMP(3)
>  )
>  WITH (
>    'connector'               = 'prometheus',
>    'metric.endpoint-url'     = 'prometheus:9090/api/v1/write',
>    'metric.name'             = 'sensorid',
>    'metric.label.keys'       = '[siteId,deviceId]',
>    'metric.sample.key'       = 'measurement',
>    'metric.sample.timestamp' = 'ts'
>  );
[INFO] Execute statement succeeded.

Flink SQL> INSERT INTO hive_catalog.prometheus.PromTable VALUES ('10221', '106', '1031', 60, TIMESTAMP '2025-05-15 08:21:00.000');
[ERROR] Could not execute SQL statement. Reason:
org.apache.flink.table.api.ValidationException: Could not find any factory for identifier 'prometheus' that implements 'org.apache.flink.table.factories.DynamicTableFactory' in the classpath.

Available factory identifiers are:

blackhole
datagen
filesystem
kafka
paimon
print
python-input-format
upsert-kafka`

nicusX commented:
@georgelza the released version DOES NOT support SQL. The support for SQL is still under development
Please do not use the PR discussion as a chat

nicusX commented on flink-connector-prometheus-request-signer-amp/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSignerFactory.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class AmazonManagedPrometheusWriteRequestSignerFactory
```
I'd be interested to see how this is used to set up the request signer declaratively in SQL

darenwkt replied:
Request signer is now a plugin (through Java Service Provider Interface (SPI)), and can be specified by user in the table config (`metric.request-signer`), for example:

```
CREATE TABLE PromTable (
  `my_metric_name` BIGINT,
  `my_label_1` BIGINT,
  `my_label_2` BIGINT,
  `sample_value` BIGINT,
  `sample_ts` TIMESTAMP(3)
)
WITH (
  'connector' = 'prometheus',
  'aws.region' = 'us-east-1',
  'metric.request-signer' = 'amazon-managed-prometheus',
  'metric.endpoint-url' = 'https://aps-workspaces.us-east-1.amazonaws.com/workspaces/abc',
  'metric.name' = 'my_metric_name',
  'metric.label.keys' = '[my_label_1,my_label_2]',
  'metric.sample.key' = 'sample_value',
  'metric.sample.timestamp' = 'sample_ts',
  'sink.batch.max-size' = '2',
  'sink.flush-buffer.size' = '1'
);
```

georgelza replied:
please explain "
  'sink.batch.max-size' = '2',
  'sink.flush-buffer.size' = '1'
 have idea what the intend is, just want to confirm...

darenwkt replied:
Hi @georgelza,

'sink.batch.max-size' determines the max size of batch records before flushing it to sink. I would recommend checking the asyncSink FLIP for more details on how it works https://cwiki.apache.org/confluence/display/FLINK/FLIP-171%3A+Async+Sink

Another helpful resource is to check other similar connector and how these config are used https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/kinesis/#sink-batch-max-size

When this PR is merged and released, we should have equivalent docs on what each of these config mean as well

nicusX commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/RowDataToPrometheusTimeSeriesConverter.java:
```diff
@@ -0,0 +1,116 @@
+                builder.withMetricName(fieldValue.getStringValue());
+            } else if (fieldName.equals(prometheusConfig.getMetricSampleKey())) {
+                sampleValue = fieldValue.getDoubleValue();
+            } else if (prometheusConfig.getLabelKeys().contains(fieldName)) {
```
LGTM
Just thinking out loud: the Label name is the field name and the value is the content.
No issue as label names must follow `[a-zA-Z_]([a-zA-Z0-9_])*` which is a subset of possible column names.
This limits the usage a bit compared to DataStream because all metrics represented by a table must have the same set of Label names. 
But I do not think anything different is possible with a table abstraction

darenwkt replied:
I may require further clarification on the question here, but my understanding is that this approach allows setting any label name actually, for example, the following label names are all valid:

```
  'metric.label.keys' = '[my_label_1, my_label_2, LABEL@123, 123%label, label(123)]'
```

nicusX replied:
Labels in Prometheus are {key, value}
you are adding the fieldName as key, and the content of the field as value
```
builder.addLabel(fieldName, fieldValue.getStringValue())
```
Which is fine. But it means all records will have the same set of Label keys, because in a table abstraction all records have the same fields by definition. Label values will be different, being the content of the fields.
As I mentioned, it's limited compared to DataStream but acceptable

nicusX replied:
"LABEL@123", "123%label" and "label(123)" would be illegal Label's key (names) for Prometheus BTW (see [specs](https://prometheus.io/docs/specs/prw/remote_write_spec/#labels)). They would be rejected by Prometheus. But the behaviour is not be inconsistent with the DataStream version where illegal label names (which are data in that case, not configurations) are rejected at runtime when you try writing. I am not suggesting a change.

darenwkt replied:
Thanks for the clarification, yes I agree, the behaviour does differ from datastream API in that the label key/name is fixed per table.

darenwkt replied:
Thanks Lorenzo, I have also added `TableAPIExample` as well, feel free to take a look. It seems like TableAPI `groupBy` is not the same as `keyBy` operation, so there might be some limitation around parallelism > 1

nicusX commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/RowDataToPrometheusTimeSeriesConverter.java:
```diff
@@ -0,0 +1,116 @@
+            } else if (fieldName.equals(prometheusConfig.getMetricSampleTimestamp())) {
+                sampleTimestamp = fieldValue.getLongValue();
+            } else {
+                throw new IllegalArgumentException(
```
It would be really nice to move this validation to the builder or the sink constructor. This would cause the job to fail when created, instead of failing when the first record is processed (putting the job in a fail-and-restart loop).
A more user friendly behaviour considering this is a misconfiguration rather than a runtime problem.

darenwkt replied:
Yes this is a good suggestion, and I can confirm that the validation are already in place, please take a look at:

- `PrometheusDynamicSinkFactory.validateMetricConfigKeys()`
- `PrometheusDynamicSinkFactory.requiredOptions()`

They are both called in sink constructor which aligns with your suggestion

nicusX reviewed (commented):
Very nice refactoring.
I have two suggestions, but nothing blocking:
1. It would be good validating whether table column and configuration are consistent at sink initialization rather than in the converter. Making the job failing on initialization when there is a misconfiguration is better than making it fail (and restart) on processing.
2. It would be nice to add an example of SQL, in form of a fake test similar to [DataStreamExample](https://github.com/apache/flink-connector-prometheus/blob/main/flink-connector-prometheus/src/test/java/org/apache/flink/connector/prometheus/sink/examples/DataStreamExample.java). It would be SQL embedded in Java but still meaningful for an end user

Also, I am unsure how the initialization of the request-signer works in SQL. Maybe another example class for that would be helpful

nicusX reviewed (approved):
Approved because my notes in my last review are not blockers

georgelza commented:
Hi  Daren

I actually think the way you defining the labels are very elegant.

G

On Mon, Jun 2, 2025 at 3:27 PM darenwkt ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/RowDataToPrometheusTimeSeriesConverter.java
> <https://github.com/apache/flink-connector-prometheus/pull/22#discussion_r2121153890>
> :
>
> > +        PrometheusTimeSeries.Builder builder = PrometheusTimeSeries.builder();
> +        Double sampleValue = null;
> +        Long sampleTimestamp = null;
> +
> +        for (int i = 0; i < fields.size(); i++) {
> +            DataTypes.Field field = fields.get(i);
> +            RowData.FieldGetter fieldGetter =
> +                    createFieldGetter(fields.get(i).getDataType().getLogicalType(), i);
> +            FieldValue fieldValue = new FieldValue(fieldGetter.getFieldOrNull(row));
> +            String fieldName = field.getName();
> +
> +            if (fieldName.equals(prometheusConfig.getMetricName())) {
> +                builder.withMetricName(fieldValue.getStringValue());
> +            } else if (fieldName.equals(prometheusConfig.getMetricSampleKey())) {
> +                sampleValue = fieldValue.getDoubleValue();
> +            } else if (prometheusConfig.getLabelKeys().contains(fieldName)) {
>
> I may require further clarification on the question here, but my
> understanding is that this approach allows setting any label name actually,
> for example, the following label names are all valid:
>
>   'metric.label.keys' = '[my_label_1, my_label_2, ***@***.***, 123%label, label(123)]'
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/flink-connector-prometheus/pull/22#discussion_r2121153890>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGZLFP7W3XYQO2NRRNNXMSL3BRGM3AVCNFSM6AAAAAB2JAJHWCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQOBYGQYTKNJUGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
You have the obligation to inform one honestly of the risk, and as a person
you are committed to educate yourself to the total risk in any activity!

Once informed & totally aware of the risk,
every fool has the right to kill or injure themselves as they see fit!

georgelza commented:
def keen to get access to the jar that implements this sink, have a blog
just about ready to implement a IoT ->kafka->flink-prometheus-Grafana
flow...

G

On Mon, Jun 2, 2025 at 7:40 PM darenwkt ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/RowDataToPrometheusTimeSeriesConverter.java
> <https://github.com/apache/flink-connector-prometheus/pull/22#discussion_r2121788185>
> :
>
> > +            DataTypes.Field field = fields.get(i);
> +            RowData.FieldGetter fieldGetter =
> +                    createFieldGetter(fields.get(i).getDataType().getLogicalType(), i);
> +            FieldValue fieldValue = new FieldValue(fieldGetter.getFieldOrNull(row));
> +            String fieldName = field.getName();
> +
> +            if (fieldName.equals(prometheusConfig.getMetricName())) {
> +                builder.withMetricName(fieldValue.getStringValue());
> +            } else if (fieldName.equals(prometheusConfig.getMetricSampleKey())) {
> +                sampleValue = fieldValue.getDoubleValue();
> +            } else if (prometheusConfig.getLabelKeys().contains(fieldName)) {
> +                builder.addLabel(fieldName, fieldValue.getStringValue());
> +            } else if (fieldName.equals(prometheusConfig.getMetricSampleTimestamp())) {
> +                sampleTimestamp = fieldValue.getLongValue();
> +            } else {
> +                throw new IllegalArgumentException(
>
> Yes this is a good suggestion, and I can confirm that the validation are
> already in place, please take a look at:
>
>    - PrometheusDynamicSinkFactory.validateMetricConfigKeys()
>    - PrometheusDynamicSinkFactory.requiredOptions()
>
> They are both called in sink constructor which aligns with your suggestion
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/flink-connector-prometheus/pull/22#discussion_r2121788185>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGZLFP4QHMAFJJKZDWJBOE33BSEB5AVCNFSM6AAAAAB2JAJHWCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQOBZGM4DKNJZGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
You have the obligation to inform one honestly of the risk, and as a person
you are committed to educate yourself to the total risk in any activity!

Once informed & totally aware of the risk,
every fool has the right to kill or injure themselves as they see fit!

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSignerFactory.java:
```diff
@@ -0,0 +1,36 @@
+
+    private String getAwsRegionFromEndpointUrl(String endpointUrl) {
+        String regex = "aps[a-zA-Z0-9-]*\\.([^.]+)\\.";
+        Pattern pattern = Pattern.compile(regex);
```
Let's make this static, instead of reinitialising on each request

darenwkt replied:
Refactored to remove this method

hlteoh37 commented on flink-connector-prometheus-request-signer-amp/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSigner.java:
```diff
@@ -98,6 +98,11 @@ private AwsCredentialsProvider getCredentialsProvider() {
+    String getAwsRegion() {
+        return awsRegion;
+    }
+
```
This is a bit of a smell, since we are exposing internal object details to test the Factory method. Can we instead test the factory method directly?

darenwkt replied:
Refactored to remove this method

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkBase.java:
```diff
@@ -0,0 +1,189 @@
+import java.net.URL;
+import java.util.Collection;
+
+public class PrometheusSinkBase<IN> extends AsyncSinkBase<IN, Types.TimeSeries> {
```
This is `@PublicEvolving`. We will be extending the public interface for the sink repository 2x.

darenwkt replied:
Removed this class as we are using 1 `PrometheusSink`, verified it already has `@PublicEvolving` annotation

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/PrometheusDynamicSink.java:
```diff
@@ -0,0 +1,185 @@
+import java.util.ServiceLoader;
+import java.util.stream.Stream;
+
+public class PrometheusDynamicSink extends AsyncDynamicTableSink<Types.TimeSeries>
```
This is `@PublicEvolving`.

darenwkt replied:
I think this should be `@Internal` instead, I have updated it but let me know if I misunderstood it

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSink.java:
```diff
@@ -18,36 +18,13 @@
-    private final String metricGroupName;
-
-    @SuppressWarnings("checkstyle:RegexpSingleline")
+public class PrometheusSink extends PrometheusSinkBase<PrometheusTimeSeries> {
```
Why don't we take the hit and simply propose a major version bump to implement a generic `PrometheusSink`? That way we remove the need to maintain two coupled public interfaces (we see this coupling from the `castToPrometheusSink()` public method in the `PrometheusSinkBase`... which is a big smell! Since the parent class creates an instance of the child class...)

darenwkt replied:
Yes, agree on a major version bump, the latest revision have removed the coupling, please take a look

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkBase.java:
```diff
@@ -0,0 +1,189 @@
+                httpUserAgent,
+                errorHandlingBehaviorConfig,
+                metricGroupName);
+    }
```
The fact that we need this is a big code smell, since the higher level abstraction generates the lower level abstraction. Given we have just released the connector, let's consider instead "fixing" the interface, upgrading the major version to remove this problem?

darenwkt replied:
Yes, agree on a major version bump, the latest revision have fixed the interface, please take a look

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/RowDataToPrometheusTimeSeriesConverter.java:
```diff
@@ -0,0 +1,116 @@
+                sampleTimestamp = fieldValue.getLongValue();
+            } else {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "%s is not a supported field, valid fields are: %s",
```
Is there a reason we don't just ignore the row in the `RowData` instead? e.g. user specifies a row in the table that is not used in the sink

darenwkt replied:
Thanks, yes I think it's better to ignore as well, that way user could specify columns that are not used by sink as well

hlteoh37 commented on flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/RowDataToPrometheusTimeSeriesConverter.java:
```diff
@@ -0,0 +1,116 @@
+            this.value = value;
+        }
+
+        private String getStringValue() {
+            return value.toString();
+        }
```
Are there any unsupported types? What happens if the field provided is a `MAP` / `ARRAY` ? 
https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/types/#list-of-data-types

darenwkt replied:
This is a good call out, we only support `org.apache.flink.table.data.StringData` type here, have added validation for this in latest revision

hlteoh37 reviewed (changes requested):
Thanks @darenwkt for the PR! 
I'm not a big fan of the way the PR is currently implemented, but I can see the reasons motivating it (non-generic interfaces). I've mainly put in larger comments, but will progress on essential "formatting" comments (e.g. @PublicEvolving etc.) after we agree on the below:

Main concerns:
1/ There is currently tight coupling between `PrometheusSinkBase` and `PrometheusSink`, which is a code smell, of wrongly directed dependency! Given this is a relatively new connector, shall we consider instead taking the hit of making a breaking public interface change, and bumping the major version, to include only 1 `PrometheusSink` with the generic type?

2/ The current type handling for RowData -> String on metric labels is a little bit too open, in my opinion. https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/types/#list-of-data-types There is a fixed set of Flink SQL data types. We should consider fixing the list of available types for fields (e.g. CHAR, VARCHAR, etc) and rejecting types that are funky like (RAW, MULTISET etc.)

darenwkt commented:
Thanks @hlteoh37 for the review! I have addressed the comments and appreciate if you could help take another look, thank you

> 1/ There is currently tight coupling between `PrometheusSinkBase` and `PrometheusSink`, which is a code smell, of wrongly directed dependency! Given this is a relatively new connector, shall we consider instead taking the hit of making a breaking public interface change, and bumping the major version, to include only 1 `PrometheusSink` with the generic type?

Yes, +1 to this, I have updated this in the latest revision

> 2/ The current type handling for RowData -> String on metric labels is a little bit too open, in my opinion. https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/types/#list-of-data-types There is a fixed set of Flink SQL data types. We should consider fixing the list of available types for fields (e.g. CHAR, VARCHAR, etc) and rejecting types that are funky like (RAW, MULTISET etc.)

This is a good callout, we should only support (CHAR, VARCHAR, STRING) for metric name and labels, I have updated this in the latest revision

georgelza commented:
Hi Darren

Is sink.batch.max-size and sink.flush-buffer.size part of the current build.

I noticed metric.name needs to map to a column in the table, I will argue
that this is not desirable... you might have multiple tables being that
will push to the prometheus store. based on the table being used the user
will want to define a text string as the name of the metrics contained...

G



On Wed, Jun 4, 2025 at 9:57 AM darenwkt ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> flink-connector-prometheus-request-signer-amp/src/main/java/org/apache/flink/connector/prometheus/sink/aws/AmazonManagedPrometheusWriteRequestSignerFactory.java
> <https://github.com/apache/flink-connector-prometheus/pull/22#discussion_r2125943268>
> :
>
> > @@ -0,0 +1,36 @@
> +package org.apache.flink.connector.prometheus.sink.aws;
> +
> +import org.apache.flink.connector.prometheus.sink.PrometheusRequestSigner;
> +import org.apache.flink.connector.prometheus.table.PrometheusConfig;
> +import org.apache.flink.connector.prometheus.table.PrometheusDynamicRequestSignerFactory;
> +
> +import java.util.regex.Matcher;
> +import java.util.regex.Pattern;
> +
> +public class AmazonManagedPrometheusWriteRequestSignerFactory
>
> Hi @georgelza <https://github.com/georgelza>,
>
> 'sink.batch.max-size' determines the max size of batch records before
> flushing it to sink. I would recommend checking the asyncSink FLIP for more
> details on how it works
> https://cwiki.apache.org/confluence/display/FLINK/FLIP-171%3A+Async+Sink
>
> Another helpful resource is to check other similar connector and how these
> config are used
> https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/kinesis/#sink-batch-max-size
>
> When this PR is merged and released, we should have equivalent docs on
> what each of these config mean as well
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/flink-connector-prometheus/pull/22#discussion_r2125943268>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGZLFPZT6VNQTEBKLASEMFD3B2RFFAVCNFSM6AAAAAB2JAJHWCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQOJVGY4TINRQHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
You have the obligation to inform one honestly of the risk, and as a person
you are committed to educate yourself to the total risk in any activity!

Once informed & totally aware of the risk,
every fool has the right to kill or injure themselves as they see fit!

darenwkt commented:
Hi @georgelza , 

> Is sink.batch.max-size and sink.flush-buffer.size part of the current build.

Can I clarify on your question please? Do you mean if the current TableAPI connector implementation allows user to configure their own `sink.batch.max-size` and `sink.flush-buffer.size`?

> I noticed metric.name needs to map to a column in the table, I will argue
that this is not desirable... you might have multiple tables being that
will push to the prometheus store. based on the table being used the user
will want to define a text string as the name of the metrics contained...

In the Datastream API connector, we allow PrometheusSink to submit to different MetricName according to what is set in TimeSeries. Hence, the TableAPI connector is consistent in this regard as well. https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/datastream/prometheus/#populating-a-prometheustimeseries

georgelza commented:
@darenwkt 
thanks for the responses.

w.r.t.   sink.batch.max-size and sink.flush-buffer.size? using it ran into the inherent rules of the one has to be bigger than the other, but when I followed that had some other errors... so eventually for the blog i simply removed both,

got it working, and it works nicely :) took some figuring out how the connector works, how to use Prometheus remote write and then also the time of the prometheus server and time of my metrics can't be to far in past (historic data) or to Marin future, time offset between prometheus server and data source TZ.

wrote a blog, using it, thats scheduled to come out tomorrow actually.
Will be watching this one, think it will have some good legs on it.
thanks for the work guys, appreciated.
